Fix autograd engine callback error propagation from device thread

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113702

The existing try-catch doesn't work because it doesn't call err.persist(). This is in contrast to the try-catch for evaluate_function which does work because it calls into python_engine's thread_on_exception which calls persist.

Calling persist on a python_error stashes the PyErr state from the thread-local PyThreadState onto the python_error object, so that when this error object is stored onto the future and passed back to the calling cpu thread, python_engine's execute try-catch can then err.restore() the error state. Finally, the python_engine's execute would re-raise so that this is re-caught by the HANDLE_TH_ERRORS macro.

Fixes https://github.com/pytorch/pytorch/issues/75750


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

soulitzer committed:
Fix queue_callback error propagation

[ghstack-poisoned]

soulitzer committed:
Update on "Fix autograd engine callback error propagation from device thread"


Fixes https://github.com/pytorch/pytorch/issues/75750


[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -53,27 +53,6 @@ PyObject* THPGradientEdgeClass = nullptr;
-}
-
-} // namespace
-
```
Not sure I can remove this for the dist engine actually, still need to investigate why.

soulitzer replied:
Not doing this anymore in this PR, left a note.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -89,7 +68,7 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
   THPObjectPtr pyInputs(PyTuple_New(static_cast<Py_ssize_t>(num_inputs)));
   if (!pyInputs)
-    throw_python_error();
+    throw python_error();
```
The idea here is that even when these functions are used during the backward, the engine is the one responsible for persisting the error. Never the user code?

soulitzer replied:
Yeah, but apparently not in all cases (in particular the distributed case). I'm reverting this change, and adding a todo in the comments to do as a follow up.

albanD reviewed (approved):
SGTM

soulitzer committed:
Update on "Fix autograd engine callback error propagation from device thread"


The existing try-catch doesn't work because it doesn't call err.persist(). This is in contrast to the try-catch for evaluate_function which does work because it calls into python_engine's thread_on_exception which calls persist.

Calling persist on a python_error stashes the PyErr state from the thread-local PyThreadState onto the python_error object, so that when this error object is stored onto the future and passed back to the calling cpu thread, python_engine's execute try-catch can then err.restore() the error state. Finally, the python_engine's execute would re-raise so that this is re-caught by the HANDLE_TH_ERRORS macro.

Fixes https://github.com/pytorch/pytorch/issues/75750


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge

albanD commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -392,8 +392,23 @@ PyObject* THPEngine_queue_callback(PyObject* self, PyObject* _callback) {
+    if (!result) {
+      // Note [ Persisting PyErr state across autograd engine threads ]
+      //
+      // Throw a python_error with the PyErr state persisted so that we don't
+      // lose the error state when returning to the cpu thread. Ordinarily, if
+      // we were calling into python during backward, e.g. executing a hook
+      // or custom autograd Function, we would be able to directly use
+      // `throw python_error()`. This is because (1) the engine has logic around
+      // evaluate_function that catches and persists errors, and (2) we init the
+      // PyThreadState on the creation of the thread so we aren't actually
+      // creating a separate PyThreadState everytime we call gil_scoped_acquire.
+      // Errors raised by queue_callback won't be caught by the try-catch around
+      // evaluate_function, so we need to replicate the persisting logic here.
```
I'm not sure about this comment.
I think we should make this note a big more general and add it to the error handling in evaluate function as well.
It should then say that:
- Since the engine is using multithreading, it must preserve the python error from the worker thread and rethrow it as-is in the calling thread.
- This is done by persisting errors in the two places where python errors can happen: evaluate function and callback.


I'm not sure what the point related to (2) is so I would suggest removing it.

soulitzer replied:
That's a good idea. I will update this note.

(2) is the reason we shouldn't need to immediately persist using something like `throw_python_error()`

soulitzer replied:
I haven't verified (2) myself, but per the comment in `PythonEngine::thread_init`
```
// Create a PyThreadState, but release the GIL. This lets
// pybind11::gil_scoped_acquire calls inside thread_main acquire the GIL
// without having to create a new PyThreadState each time.
```

albanD replied:
Sure but that comment is just about speeding up gil acquisition by not having to re-allocate the state everytime. It doens't have anything to do with error handling AFAIK

soulitzer replied:
Since the PyError indicators are stored on the state, I think that would go away if we need to re-allocated it. More context here as well https://github.com/pytorch/pytorch/pull/34845

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -56,16 +56,12 @@ PyObject* THPGradientEdgeClass = nullptr;
+// create a shared PyThreadState upon initialization, we should be able
+// to rely on the engine to persist for us. But, removing this seems to
+// cause the following test to fail:
+// python test/distributed/rpc/test_tensorpipe_agent.py -k test_backward_autograd_engine_error
```
You mean removing the use of this function right? Not this function itself?
Is the distributed engine missing a python counterpart that would be persisting the errors in general?
If so I would expect that any error in a hook during backward would hard crash it?

soulitzer replied:
> You mean removing the use of this function right? Not this function itself?

It's only used in this file AFAIK, so both?

> Is the distributed engine missing a python counterpart that would be persisting the errors in general?

It holds an instance of the non-dist engine, constructed via get_default_engine, so presumably it would correctly create the right one (haven't verified).

> If so I would expect that any error in a hook during backward would hard crash it?

Good point, I would expect this to be the case (haven't verified).

soulitzer replied:
> You mean removing the use of this function right? Not this function itself?

Clarified this comment. 

I've updated the PR to fix the dist engine hook case.

soulitzer committed:
Update on "Fix autograd engine callback error propagation from device thread"


The existing try-catch doesn't work because it doesn't call err.persist(). This is in contrast to the try-catch for evaluate_function which does work because it calls into python_engine's thread_on_exception which calls persist.

Calling persist on a python_error stashes the PyErr state from the thread-local PyThreadState onto the python_error object, so that when this error object is stored onto the future and passed back to the calling cpu thread, python_engine's execute try-catch can then err.restore() the error state. Finally, the python_engine's execute would re-raise so that this is re-caught by the HANDLE_TH_ERRORS macro.

Fixes https://github.com/pytorch/pytorch/issues/75750


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Fix autograd engine callback error propagation from device thread"


The existing try-catch doesn't work because it doesn't call err.persist(). This is in contrast to the try-catch for evaluate_function which does work because it calls into python_engine's thread_on_exception which calls persist.

Calling persist on a python_error stashes the PyErr state from the thread-local PyThreadState onto the python_error object, so that when this error object is stored onto the future and passed back to the calling cpu thread, python_engine's execute try-catch can then err.restore() the error state. Finally, the python_engine's execute would re-raise so that this is re-caught by the HANDLE_TH_ERRORS macro.

Fixes https://github.com/pytorch/pytorch/issues/75750


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -26,6 +26,17 @@ static void check_single_result(
+}
+
+} // namespace
+
```
Maybe I should dedupe this function, but putting it in python_variable.h

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -575,6 +575,7 @@ auto Engine::thread_main(const std::shared_ptr<GraphTask>& graph_task) -> void {
                 local_graph_task->cpu_ready_queue_);
           }
         } catch (std::exception& e) {
+          // See Note [ Persisting PyErr state across autograd engine threads ]
```
I meant to put that in the implementation of thread_on_exception in the python_engine: the place where we do persist the error)

soulitzer replied:
Oh I was thinking that putting it here would be more visible since when debugging this I didn't even realize that the python engine had overridden this function! Agreed that the implementation is probably the more canonical place though. Going to put in both places.

albanD commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -392,8 +392,25 @@ PyObject* THPEngine_queue_callback(PyObject* self, PyObject* _callback) {
+      // persisting the error in the two places that can encounter Python
+      // errors: (1) evaluate function and (2) queued callbacks.
+      //
+      // TODO: the engine is not actually responsible for persisting the error
```
So all of these can be removed?

soulitzer replied:
Errr I should this comment more precise. The engine is not responsible for persisting the error particularly in the custom Function and backward hooks cases today, but I haven't looked at any other cases (torch dispatch, PyInterpreter cases) so if we want to remove the persist in the engine, those cases also need to be checked that they aren't being regressed. 

I'm not sure its entirely worth it to look into more of these cases if we already believe that the best solution is for persist to be handled in the engine. Like if someone wants to fix the DistEngine torch dispatch error propagation for example (I don't know if its actually broken), I'd prefer that they fix the issue systematically rather than have to exhaustively insert the persist at every individual call site.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -50,22 +50,21 @@ PyObject* THPGradientEdgeClass = nullptr;
   if (!(condition)) {                      \
     THPUtils_setError(__VA_ARGS__);        \
-    throw python_error();                  \
+    throw_python_error();                  \
```
I'm confused... Isn't your comment above saying that only the engine should do persist and the user code never has to do it?

soulitzer committed:
Update on "Fix autograd engine callback error propagation from device thread"


The existing try-catch doesn't work because it doesn't call err.persist(). This is in contrast to the try-catch for evaluate_function which does work because it calls into python_engine's thread_on_exception which calls persist.

Calling persist on a python_error stashes the PyErr state from the thread-local PyThreadState onto the python_error object, so that when this error object is stored onto the future and passed back to the calling cpu thread, python_engine's execute try-catch can then err.restore() the error state. Finally, the python_engine's execute would re-raise so that this is re-caught by the HANDLE_TH_ERRORS macro.

Fixes https://github.com/pytorch/pytorch/issues/75750


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
